orderer/consensus/hlmirbft/chain.go

```diff
@@ -1056,146 +1094,196 @@ func (c *Chain) triggerCatchup(sn *raftpb.Snapshot) {
 			c.writeConfigBlock(configBlock)
 		} else {
-      
-			envs = append(envs,env)
-      
+
+			envs = append(envs, env)
+
 		}
 
-}
-if len(envs)!=0 {
+	}
+	if len(envs) != 0 {
 		block := c.CreateBlock(envs)
 		c.writeBlock(block)
-}
-
+	}
 
 	return nil
 }
 
-
 //JIRA FLY2-48 proposed changes:Write block in accordance with the sequence number
 func (c *Chain) Apply(batch *msgs.QEntry) error {
 	c.pendingBatches[batch.SeqNo] = batch
-	index := 0  // Review comment change to rpelace append by index.
+	index := 0 // Review comment change to rpelace append by index.
 	seqNumbers := make([]uint64, len(c.pendingBatches))
 	for k := range c.pendingBatches {
 		seqNumbers[index] = k
 		index++
 	}
 	sort.SliceStable(seqNumbers, func(i, j int) bool { return seqNumbers[i] < seqNumbers[j] })
-	for i:=0;i<len(seqNumbers);i++{
-			if c.Node.LastCommittedSeqNo+1 != seqNumbers[i] {
-				break
-			}
-			err := c.processBatch(c.pendingBatches[seqNumbers[i]])
-			if err != nil {
-			    return errors.WithMessage(err, "Batch Processing Error")
-			}
-			delete(c.pendingBatches, seqNumbers[i])
-			c.Node.LastCommittedSeqNo++
+	for i := 0; i < len(seqNumbers); i++ {
+		if c.Node.LastCommittedSeqNo+1 != seqNumbers[i] {
+			break
+		}
+		err := c.processBatch(c.pendingBatches[seqNumbers[i]])
+		if err != nil {
+			return errors.WithMessage(err, "Batch Processing Error")
+		}
+		delete(c.pendingBatches, seqNumbers[i])
+		c.Node.LastCommittedSeqNo++
 	}
 
 	return nil
 }
+
 //FLY2-48 proposed changes
 //	- create Blocks
-func (c *Chain ) CreateBlock(envs []*common.Envelope) *common.Block {
+func (c *Chain) CreateBlock(envs []*common.Envelope) *common.Block {
+
 	bc := &blockCreator{
 		hash:   protoutil.BlockHeaderHash(c.lastBlock.Header),
+		//change
 		number: c.lastBlock.Header.Number,
 		logger: c.logger,
 	}
 	return bc.createNextBlock(envs)
 }
 
+//JIRA FLY2-106 check if pending batch list is empty
+func (c *Chain) isPendingBatchesEmpty() bool {
+	if len(c.pendingBatches) == 0 {
+		return true
+	}
+	return false
+}
+
+//JIRA FLY2-106 sleep till pending batch list is empty
+func (c *Chain) waitForPendingBatchCommits() {
+	isEmpty := c.isPendingBatchesEmpty()
+	for !isEmpty {
+		time.Sleep(400 * time.Millisecond)
+		isEmpty = c.isPendingBatchesEmpty()
+	}
+}
 
 //JIRA FLY2-66 proposed changes:Implemented the Snap Function
 func (c *Chain) Snap(networkConfig *msgs.NetworkState_Config, clientsState []*msgs.NetworkState_Client) ([]byte, []*msgs.Reconfiguration, error) {
-	
-	pr := c.Node.PendingReconfigurations
-	
-	c.Node.PendingReconfigurations = nil
 
-	data, err := proto.Marshal(&msgs.NetworkState{
+	var pr []*msgs.Reconfiguration
+	var newNetworkConfig *msgs.NetworkState_Config
+
+	if len(c.Node.PendingReconfigurations) == 0 {
+		pr = append(pr, []*msgs.Reconfiguration{nil,nil}...)
+	}
+	pr = c.Node.PendingReconfigurations[0]
+	c.Node.PendingReconfigurations = PopReconfiguration(c.Node.PendingReconfigurations)
```

Comment:
The popping of a reconfiguration should occur after the `newNetworkConfig` has been equated with `networkConfig`. Moreover this should occur after `Snap` returns, so place it execution inside a defer statement https://tour.golang.org/flowcontrol/12

Reply:
This is so that a pending reconfiguration is returned twice - once to notify Mir of the new configuration and twice to force an epoch change that includes this new configuration.

Reply:
@harrymknight  the issue has fixed